CanvasRenderingContext2D::drawImage() doesn't work correctly with transform when copying from self

shallawa opened the pull request:
#### aa025e46a961480b60e856a371d639bd874b74a7
<pre>
CanvasRenderingContext2D::drawImage() doesn&apos;t work correctly with transform when copying from self
<a href="https://bugs.webkit.org/show_bug.cgi?id=255208">https://bugs.webkit.org/show_bug.cgi?id=255208</a>
rdar://107844963

Reviewed by Kimmo Kinnunen.

drawImage() creates a temporary ImageBuffer to get the srcRect of the canvas.
This ImageBuffer is not for drawing. It should not inherit the current scaleFactor
and it should just hold the pixels of the canvas in the non-scaled coordinates. We
should use GraphicsContext::createImageBuffer() to create the temporary ImageBuffer.

* LayoutTests/fast/canvas/canvas-self-drawImage-scaled-context-expected.html: Added.
* LayoutTests/fast/canvas/canvas-self-drawImage-scaled-context.html: Added.
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::drawImage):

Canonical link: <a href="https://commits.webkit.org/262841@main">https://commits.webkit.org/262841@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d7682e93c450a2819f60ebd0814c95c0b64c33fd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2744 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2743 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2890 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4153 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3107 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2709 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2888 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2842 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2414 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2770 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3201 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2454 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3923 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/695 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2439 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2295 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2451 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2490 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3671 "261 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2824 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2269 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2487 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2442 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/685 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2480 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2637 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on LayoutTests/fast/canvas/canvas-self-drawImage-scaled-context.html:
```diff
@@ -0,0 +1,33 @@
+            ctx.restore();
+        }
+
+        for (let i = 0; i < 4; ++i)
```
If I understand correctly, each individual iteration here will just overwrite the previous? And also that single iteration would reveal the bug?

I'd prefer test to be of form:
Reference image::
- draw a triangle to the left side

Test image:
- draw a triangle to the right side
- imageSmoothingEnabled  = false
- move the image from right to left via drawImage+global composite copy operation

When the test image is a triangle, you would more easily catch errors where the scale factor of the copy differs from the scale factor of the source and the destination. As an example, the patch puts up explicit "1" as the scale factor. So then it'd be simpler to still believe that is correct, as the test would prove it being the case.

shallawa committed:
CanvasRenderingContext2D::drawImage() doesn't work correctly with transform when copying from self
https://bugs.webkit.org/show_bug.cgi?id=255208
rdar://107844963

Reviewed by Kimmo Kinnunen.

drawImage() creates a temporary ImageBuffer to get the srcRect of the canvas.
This ImageBuffer is not for drawing. It should not inherit the current scaleFactor
and it should just hold the pixels of the canvas in the non-scaled coordinates. We
should use GraphicsContext::createImageBuffer() to create the temporary ImageBuffer.

* LayoutTests/fast/canvas/canvas-self-drawImage-scaled-context-expected.html: Added.
* LayoutTests/fast/canvas/canvas-self-drawImage-scaled-context.html: Added.
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::drawImage):

Canonical link: https://commits.webkit.org/262841@main
